### ISSUE TITLE
[BUG] Package ML-PACE triggers error when building with Kokkos_ENABLE_CUDA

### ISSUE THREAD

**Issue by ocaisa** [lammps/lammps#3353]
**Summary**

I'm trying to create updated recipes for EasyBuild of LAMMPS, with and without GPU support. The CPU-only version builds ok, the GPU-enabled version (using Kokkos) errors out in the ML-PACE package with errors such as
```
/tmp/eb-z7lua0g2/tmpxft_002d2661_00000000-6_binary.cudafe1.cpp:1:3: error: style of line directive is a GCC extension
    1 | # 1 "/project/60005/easybuild/build/LAMMPS/23Jun2022/foss-2021a-CUDA-kokkos/easybuild_obj/lammps-user-pace-v.2021.10.25.fix2/yaml-cpp/src/binary.cpp"
      |   ^
/project/60005/easybuild/build/LAMMPS/23Jun2022/foss-2021a-CUDA-kokkos/easybuild_obj/lammps-user-pace-v.2021.10.25.fix2/yaml-cpp/src/binary.cpp:2:3: error: style of line directive is a GCC extension
```
Removing ML-PACE from the list of packages allows the build to succeed.

**LAMMPS Version and Platform**

Version: 23Jun2022
Platform: AMD Zen2 with NVIDIA V100 GPU, Rocky Linux 8, EasyBuild recipe for LAMMPS

**Expected Behavior**

GPU build succeeds as the CPU build did with the same packages.

**Actual Behavior**

GPU-enabled build fails during the ML-PACE package, disabling this package allowed the build to succeed

**Steps to Reproduce**

The (failing) configure command used was
```
cmake -DCMAKE_INSTALL_PREFIX=/project/def-sponsor00/easybuild/software/LAMMPS/23Jun2022-foss-2021a-CUDA-kokkos\
 -DCMAKE_BUILD_TYPE=Release -DCMAKE_SYSROOT=/cvmfs/pilot.eessi-hpc.org/versions/2021.12/compat/linux/x86_64\
 -DCMAKE_C_COMPILER='/cvmfs/pilot.eessi-hpc.org/versions/2021.12/software/linux/x86_64/amd/zen2/software/OpenMPI/4.1.1-GCC-10.3.0/bin/mpicc'\
 -DCMAKE_C_FLAGS='-O2 -ftree-vectorize -march=native -fno-math-errno -fopenmp'\
 -DCMAKE_CXX_COMPILER='/cvmfs/pilot.eessi-hpc.org/versions/2021.12/software/linux/x86_64/amd/zen2/software/OpenMPI/4.1.1-GCC-10.3.0/bin/mpicxx'\
 -DCMAKE_CXX_FLAGS='-O2 -ftree-vectorize -march=native -fno-math-errno -fopenmp'\
 -DCMAKE_Fortran_COMPILER='/cvmfs/pilot.eessi-hpc.org/versions/2021.12/software/linux/x86_64/amd/zen2/software/OpenMPI/4.1.1-GCC-10.3.0/bin/mpifort'\
 -DCMAKE_Fortran_FLAGS='-O2 -ftree-vectorize -march=native -fno-math-errno -fopenmp'\
 -DCMAKE_SKIP_RPATH=ON -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_FIND_USE_PACKAGE_REGISTRY=FALSE\
 -DBOOST_ROOT=/project/def-sponsor00/easybuild/software/Boost/1.76.0-GCC-10.3.0 -DBoost_NO_SYSTEM_PATHS=ON\
  -DBUILD_TOOLS=on  -DBUILD_DOC=off  -DWITH_GZIP=yes  -DWITH_PNG=yes  -DWITH_JPEG=yes\
  -DDOWNLOAD_EIGEN3=no  \
-DEIGEN3_INCLUDE_DIR=/cvmfs/pilot.eessi-hpc.org/versions/2021.12/software/linux/x86_64/amd/zen2/software/Eigen/3.3.9-GCCcore-10.3.0/include/Eigen\
  -DEigen3_DIR=/cvmfs/pilot.eessi-hpc.org/versions/2021.12/software/linux/x86_64/amd/zen2/software/Eigen/3.3.9-GCCcore-10.3.0/share/eigen3/cmake/\
  -DPKG_ASPHERE=on  -DPKG_ATC=on  -DPKG_AWPMD=on  -DPKG_BOCS=on  -DPKG_BODY=on  -DPKG_BPM=on\
  -DPKG_BROWNIAN=on  -DPKG_CG-DNA=on  -DPKG_CG-SDK=on  -DPKG_CLASS2=on  -DPKG_COLLOID=on\
  -DPKG_COLVARS=on  -DPKG_COMPRESS=on  -DPKG_CORESHELL=on  -DPKG_DIELECTRIC=on  -DPKG_DIFFRACTION=on\
  -DPKG_DIPOLE=on  -DPKG_DPD-BASIC=on  -DPKG_DPD-MESO=on  -DPKG_DPD-REACT=on  -DPKG_DPD-SMOOTH=on\
  -DPKG_DRUDE=on  -DPKG_EFF=on  -DPKG_ELECTRODE=on  -DPKG_EXTRA-COMPUTE=on  -DPKG_EXTRA-DUMP=on\
  -DPKG_EXTRA-FIX=on  -DPKG_EXTRA-MOLECULE=on  -DPKG_EXTRA-PAIR=on  -DPKG_FEP=on  -DPKG_GRANULAR=on\
  -DPKG_H5MD=on  -DPKG_INTERLAYER=on  -DPKG_KIM=on  -DPKG_KSPACE=on  -DPKG_LATBOLTZ=on \
 -DPKG_MANIFOLD=on  -DPKG_MACHDYN=on  -DPKG_MANYBODY=on  -DPKG_MC=on  -DPKG_MEAM=on  -DPKG_MGPT=on\
  -DPKG_MISC=on  -DPKG_ML-IAP=on  -DPKG_ML-PACE=on  -DPKG_ML-RANN=on  -DPKG_ML-SNAP=on  -DPKG_MOFFF=on\
  -DPKG_MOLECULE=on  -DPKG_MOLFILE=on  -DPKG_MPIIO=on  -DPKG_NETCDF=on  -DPKG_OPENMP=on\
  -DPKG_ORIENT=on  -DPKG_PERI=on  -DPKG_PHONON=on  -DPKG_PLUGIN=on  -DPKG_PLUMED=on  -DPKG_POEMS=on\
  -DPKG_PTM=on  -DPKG_PYTHON=on  -DPKG_QEQ=on  -DPKG_QTB=on  -DPKG_REACTION=on  -DPKG_REAXFF=on\
  -DPKG_REPLICA=on  -DPKG_RIGID=on  -DPKG_SCAFACOS=on  -DPKG_SHOCK=on  -DPKG_SMTBQ=on  -DPKG_SPH=on\
  -DPKG_SPIN=on  -DPKG_SRD=on  -DPKG_TALLY=on  -DPKG_UEF=on  -DPKG_VORONOI=on  -DPKG_YAFF=on\
  -DPKG_OPT=on  -DBUILD_MPI=yes  -DBUILD_OMP=yes  -DPKG_OPENMP=on  -DFFT=FFTW3  -DFFT_PACK=array\
  -DPKG_KOKKOS=on  -DKokkos_ENABLE_OPENMP=yes  -DKokkos_ENABLE_CUDA=yes\
  -DCMAKE_CXX_COMPILER="/project/60005/easybuild/build/LAMMPS/23Jun2022/foss-2021a-CUDA-kokkos/lammps-stable_23Jun2022/lib/kokkos/bin/nvcc_wrapper" \
 -DCMAKE_CXX_FLAGS="-ccbin $CXX $CXXFLAGS"  -DKokkos_ARCH_ZEN2=yes  -DKokkos_ARCH_AMPERE80=yes\
  -DCMAKE_INSTALL_LIBDIR=lib \
 -DPython_EXECUTABLE=/cvmfs/pilot.eessi-hpc.org/versions/2021.12/software/linux/x86_64/amd/zen2/software/Python/3.9.5-GCCcore-10.3.0/bin/python \
 -DBUILD_SHARED_LIBS=ON \
 /project/60005/easybuild/build/LAMMPS/23Jun2022/foss-2021a-CUDA-kokkos/lammps-stable_23Jun2022/cmake
```

**Further Information, Files, and Links**

**Comment by akohlmey** [comment]
@ocaisa This is not really a LAMMPS bug and not something *we* can fix as the error happens in the yaml-cpp package that is included with the ML-PACE package library, not LAMMPS itself and one can argue that the real bug is with nvcc and its nature as a compiler-chimera or with Kokkos' requirement to compile all code when configured for CUDA with nvcc (and thus cause all kinds of errors if the host compiler uses/supports code constructs nvcc doesn't understand or know how to handle).

**Comment by akohlmey** [comment]
Perhaps @stanmoore1 can tell you how exactly he compiled this without running into the same issue.

**Comment by ocaisa** [comment]
Others have indeed run into this: https://github.com/jbeder/yaml-cpp/issues/831

**Comment by ocaisa** [comment]
Ah, interesting, the nvcc wrapper already should fix this (https://github.com/lammps/lammps/blob/develop/lib/kokkos/bin/nvcc_wrapper#L312) but may not catch `-pedantic-errors`

**Comment by akohlmey** [comment]
Have you tried compiling without `-DCMAKE_CXX_COMPILER=`? I had removed it from the `kokkos-cuda` preset since the current versions of Kokkos do not require setting it explicitly anymore and then compilation of "external" packages would not even be passed through nvcc via the nvcc_wrapper script resulting in an significant compilation speedup.

**Comment by ocaisa** [comment]
Ah, I wasn't aware of that, I will try it, thanks.

The CMake advice at https://docs.lammps.org/Build_extras.html#kokkos does still say it should be set...

**Comment by ocaisa** [comment]
For the record, patching https://github.com/lammps/lammps/blob/develop/lib/kokkos/bin/nvcc_wrapper#L312 to
```
-pedantic|-pedantic-errors|-Wpedantic|-ansi)
```
did seem to fix the problem

**Comment by akohlmey** [comment]
> Ah, I wasn't aware of that, I will try it, thanks.
> 
> The CMake advice at https://docs.lammps.org/Build_extras.html#kokkos does still say it should be set...

This was a new feature at the time and we were being conservative in the documentation. I'll probably only update the docs for the next development cycle and incorporate your patch to nvcc_wrapper into PR #3324.

If you are building the stable version, you probably want to incorporate the accumulated patches from that PR (or wait until we release the first update to the stable release). The number of backported bugfixes is rather small so far, though.

**Comment by ocaisa** [comment]
I see the fix included in #3324 so closing this, thanks for the help!

**Comment by ocaisa** [comment]
(BTW the wrapper patch is also merged into Kokkos via https://github.com/kokkos/kokkos/pull/5235)
